crates/core/tedge_api/src/partial_entity_store.rs

```diff
@@ -0,0 +1,167 @@
+use crate::entity_store::EntityRegistrationMessage;
+use crate::mqtt_topics::Channel;
+use crate::mqtt_topics::EntityTopicId;
+use crate::ring_buffer::RingBuffer;
+use mqtt_channel::Message as MqttMessage;
+use std::collections::HashMap;
+
+const EARLY_MESSAGE_BUFFER_SIZE: usize = 5;
+
+/// A store for all the entities for which data messages are received before
+/// its registration message itself is received.
+/// It also stores all the child device registration messages received before
+/// their parents themselves are registered, including their data.
+pub struct PartialEntityStore {
+    main_device_tid: EntityTopicId,
+    // This orphans map is keyed by the unregistered parent topic id to their children
+    orphans: HashMap<EntityTopicId, Vec<EntityTopicId>>,
+    entities: HashMap<EntityTopicId, PartialEntityCache>,
```

Comment:
> Having a cache per unregistered entity makes this partial entity store fragile when facing messages on random entities.

I didn't get that fragility concern.

One reason why I kept it per entity is to make the retrieval of all messages for a given entity easier. I could create a single buffer for all entities as well, despite the need to search the entire buffer to retrieve messages for each entity. But, since these buffers are gonna be small, that's not a big deal. But, this impl with map seemed much easier to handle.

Reply:
>> Having a cache per unregistered entity makes this partial entity store fragile when facing messages on random entities.

> I didn't get that fragility concern.

As there are an unlimited number of buffers, the cache is actually unlimited even if each individual cache is limited.

> One reason why I kept it per entity is to make the retrieval of all messages for a given entity easier. I could create a single buffer for all entities as well, despite the need to search the entire buffer to retrieve messages for each entity. But, since these buffers are gonna be small, that's not a big deal. But, this impl with map seemed much easier to handle.

Having to go through a list of pending messages to filter out those that can be applied after a new registration should not be neither too complex nor inefficient as this cache if for less than one hundred of messages.